[Console] Handle calls to mb_ functions with non string arguments

Yopai opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #40200
| License       | MIT
| Doc PR        | no

In PHP8.1, a number of functions who were accepting null arguments will only accept
string ones.
(see https://wiki.php.net/rfc/deprecate_null_to_scalar_internal_arg)

In the polyfill, mb_* functions are already declared with a strict type checking of "string".

Therefore, it is necessary to get rid of the use of non string arguments when calling mb_* functions,
so that it won't break when either using the polyfill,or future php8 versions.

In every call where the argument may not be a string, this commit enforces the string type of the argument (with transtyping)

--- For reviewers
* I generally don't like transtyping, but found it was the more "secure" way (on a non-BC point of view) here.
Specially in Console/Helper/Table.php, where $cell can be an object (there are 2 "$cell instanceof ... tests)
However, where the argument can already be either null or string (and not anything else), there may a beter approach ?

* It's the first time I send a PR on symfony, so don't hesitate pointing me to thinks I've forgotten to done.

Yopai committed:
In calls to mb_ functions, silently transform arg into string

In PHP8, a number of functions who were accepting null arguments will only accept
string ones.

In the polyfill, mb_* functions are declared with a trict type checking of "string".

Therefore, we deprecate the use of non string arguments, so that it won't break when either using the polyfill,
or future php8 versions.

derrabus commented on src/Symfony/Component/Console/Helper/Helper.php:
```diff
@@ -47,6 +47,9 @@ public function getHelperSet()
      */
     public static function strlen($string)
     {
+        if ( ! \is_string($string)) {
+            $string = (string)$string;
+        }
```
```suggestion
        $string = (string) $string;
```

There are more places than this. We don't need to type-check here because a string cast on a string should be a no-op anyway.

derrabus commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -512,9 +512,11 @@ private function renderCell(array $row, int $column, string $cellFormat): string
-        // str_pad won't work properly with multi-byte strings, we need to fix the padding
-        if (false !== $encoding = mb_detect_encoding($cell, null, true)) {
-            $width += \strlen($cell) - mb_strwidth($cell, $encoding);
+        if ( ! is_null($cell)) {
+            // str_pad won't work properly with multi-byte strings, we need to fix the padding
+            if (false !== $encoding = mb_detect_encoding((string)$cell, null, true)) {
+                $width += \strlen((string)$cell) - mb_strwidth((string)$cell, $encoding);
+            }
```
```suggestion
        // str_pad won't work properly with multi-byte strings, we need to fix the padding
        if (null !== $cell && false !== $encoding = mb_detect_encoding((string)$cell, null, true)) {
            $width += \strlen((string)$cell) - mb_strwidth((string)$cell, $encoding);
```

derrabus commented:
Please also have a look at the CS errors fabbot reported.

Yopai commented:
Thanks.
I fixed not-useful tests, and fix CS for what I modified.
And then I fixed use of shorttag, that fabbot indicated too.
(I pushed the commits again, is there anything else I have to do ?)

derrabus commented on src/Symfony/Component/Mime/Test/Constraint/EmailHtmlBodyContains.php:
```diff
@@ -43,7 +43,7 @@ protected function matches($message): bool
         }
 
-        return false !== mb_strpos($message->getHtmlBody(), $this->expectedText);
+        return false !== mb_strpos((string) ($message->getHtmlBody()), $this->expectedText);
```
```suggestion
        return false !== mb_strpos((string) $message->getHtmlBody(), $this->expectedText);
```

derrabus commented on src/Symfony/Component/Mime/Test/Constraint/EmailTextBodyContains.php:
```diff
@@ -43,7 +43,7 @@ protected function matches($message): bool
         }
 
-        return false !== mb_strpos($message->getTextBody(), $this->expectedText);
+        return false !== mb_strpos((string) ($message->getTextBody()), $this->expectedText);
```
```suggestion
        return false !== mb_strpos((string) $message->getTextBody(), $this->expectedText);
```

derrabus reviewed (changes requested):
Sorry, I should've been more precise. You don't need to fix fabbot errors you haven't caused. In this case, the short tags were a false positive. Can you restore them?

Yopai commented:
No worry (that's why I made it in a separate commit :) )
Done !

derrabus commented on src/Symfony/Component/Form/Tests/Util/StringUtilTest.php:
```diff
@@ -34,7 +34,7 @@ public function testTrimUtf8Separators($hex)
 
         // Convert UCS-2BE to UTF-8
-        $symbol = mb_convert_encoding($binary, 'UTF-8', 'UCS-2BE');
+        $symbol = mb_convert_encoding((string) $binary, 'UTF-8', 'UCS-2BE');
```
Do we need this change? `pack()` should always return a string, shouldn't it?

Yopai replied:
From the doc : it can return false on error. As I'm not sure if/how such an error is handled, I've prefered transtyping it too. But I can remove it if you think it's unnecessary.

derrabus replied:
> it can return false on error

On PHP 8, we should get a `ValueError` instead. So on PHP 8.1 we can be certain that `$binary` is a string. So yes, please revert this line.

Yopai replied:
Done.

nicolas-grekas commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -513,8 +513,8 @@ private function renderCell(array $row, int $column, string $cellFormat): string
         // str_pad won't work properly with multi-byte strings, we need to fix the padding
-        if (false !== $encoding = mb_detect_encoding($cell, null, true)) {
-            $width += \strlen($cell) - mb_strwidth($cell, $encoding);
+        if (null !== $cell && false !== $encoding = mb_detect_encoding((string) $cell, null, true)) {
```
`$cell` can't be null here
if the cast to string is needed, let's make it happen once

Yopai replied:
Done.

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/BinaryFileResponse.php:
```diff
@@ -155,6 +155,8 @@ public function setAutoEtag()
     {
+        $filename = (string) $filename;
+
         if ('' === $filename) {
```
if ('' === ($filename ?? '')) {

Yopai replied:
Why not simplier `if (!$filename)` ?

Or `if (empty($filename))` ?

nicolas-grekas replied:
it wouldn't be the same, as `'0'` is false|empty

Yopai replied:
(We could also assume that filename is necessarily a string, as phpdoc doesn't allow it to be null, is it right ?)

nicolas-grekas replied:
I believe so yes

Yopai replied:
Ok, I'll remove the typecast.

nicolas-grekas commented on src/Symfony/Component/Mime/Test/Constraint/EmailHtmlBodyContains.php:
```diff
@@ -43,7 +43,7 @@ protected function matches($message): bool
         }
 
-        return false !== mb_strpos($message->getHtmlBody(), $this->expectedText);
+        return false !== mb_strpos((string) $message->getHtmlBody(), $this->expectedText);
```
can't happen, only RawMessage is allowed

Yopai replied:
The type hinting of getHtmlBody() says "null|resource|string".

In fact, looking for the type of $message (assuming the phpdoc type is respected - even if it is not strictly enforced with type hinting), we can assume :
- that $message is an "instanceof" RawMessage 
- but not directly a RawMessage or a Message.
Nothing ensures its getText/HtmlBody method returns a string. Nor it even exists.

Maybe this classes would require a refactoring, with creation of an interface ?

Yopai replied:
(but it's outside the scope of this PR)

nicolas-grekas replied:
I'm for fixing things that are actual issues.
If you can add a test case for this, I'm fine :)
But then, we should also handle the case where a resource is returned.

Yopai replied:
For the case a resource is returned, the difference is: it won't break _more_ with PHP8.1 or the polyfill, because it will _already_ break.

Again: I checked (at your demand!) what could be checked, i.e when the type error could occur with PHP8.1 although it doesn't occur with previous PHP versions. Even if someone "misuse" the code, like calling addRow with an array of whatever he wants.

All my edits had to reduce inconsistency between current and future php versions / polyfill. I'm just sorry it wouldn't have been understood like that.

nicolas-grekas commented on src/Symfony/Component/Mime/Test/Constraint/EmailTextBodyContains.php:
```diff
@@ -43,7 +43,7 @@ protected function matches($message): bool
         }
 
-        return false !== mb_strpos($message->getTextBody(), $this->expectedText);
+        return false !== mb_strpos((string) $message->getTextBody(), $this->expectedText);
```
can't happen

nicolas-grekas commented on src/Symfony/Component/Translation/Dumper/IcuResFileDumper.php:
```diff
@@ -42,6 +42,8 @@ public function formatCatalogue(MessageCatalogue $messages, $domain, array $opti
         $keyTop = $this->getPosition($data);
 
         foreach ($messages->all($domain) as $source => $target) {
+            $target = (string) $target;
```
can really the target be null?

Yopai replied:
Strictly speaking, there's no type hinting on MessageCatalogue::$messages
But... Okay, there are supposed to be strings.
Fixed

nicolas-grekas commented on src/Symfony/Component/VarDumper/Dumper/CliDumper.php:
```diff
@@ -185,6 +185,8 @@ public function dumpScalar(Cursor $cursor, $type, $value)
      */
     public function dumpString(Cursor $cursor, $str, $bin, $cut)
     {
+        $str = (string) $str;
```
cannot be null

Yopai replied:
You're right, didn't see the phpdoc on interface.
Fixed.

nicolas-grekas commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -513,8 +513,9 @@ private function renderCell(array $row, int $column, string $cellFormat): string
         // str_pad won't work properly with multi-byte strings, we need to fix the padding
-        if (false !== $encoding = mb_detect_encoding($cell, null, true)) {
-            $width += \strlen($cell) - mb_strwidth($cell, $encoding);
+        $cellstr = (string) $cell;
```
$cell as name would save renaming below,thus save merge conflicts, please keep the name

Yopai replied:
I would have done it if `$cell` wasn't used below :
`$cell instanceof TableSeparator`

Do you prefer that I store `$cell instanceof TableSeparator` in a variale, so I can rewrite this as

	$isTableSeparator = ($cell instanceof TableSeparator)

	$cell = (string) $cell;
	if (false !== $encoding = mb_detect_encoding($cell, null, true)) {
		$width += \strlen($cell) - mb_strwidth($cell, $encoding);
	}

	$style = $this->getColumnStyle($column);

	if ($isTableSeparator) {
		return sprintf($style->getBorderFormat(), str_repeat($style->getBorderChars()[2], $width));
	}

nicolas-grekas replied:
Oh, I missed this.
But: is this needed at all? $cell cannot be null here.

Yopai replied:
I must have missed something, I wonder how you can be sure ?
In fact, addRow is public, and can take an array of anything, since nothing enforces the type of the elements. So, IMHO, in renderRow, you can't assume anything on the type of elements of row. And, therefore, you can't assume anything on the type of $cell.

BTW, even if it was right, the aim of this PR is to avoid errors when upgrading to PHP8.1 (or to the usage of the polyfill) by transtyping _before_ the call to the mb_ function, rather than relying on the implict cast done until php8.0

I originally had the problem with a null value, you're right, but the problem is not related with the value being null, it is related with the value being anything else than a string.
Arguments of mb_* functions _must_ be a string (in PHP8.1, and therefore it has been coded in the polyfill). Passing them an object won't be good.

So :
- In any way, it can be an object (since code is testing for it)
- It comes from a public method, with no type checking - not even in phpdoc -, so I don't see how you can assume anything about its type.

(All that being said, I am perfectly conscious about the work you provide, and the very good knowledge of this code you have. That's why you're reviewers and I'm just a first-time request-puller :) )

nicolas-grekas replied:
Passing them a Stringable object will continue to work, isn't it?
That's how this piece of code is designed to me, and it's fine as is.

Yopai replied:
AFAIK, the prototype of the mb_ functions is marked (in the php documentation) as "string", not "string|Stringable"

And if I undesrtand correctly https://php.watch/versions/8.0/stringable, this has been implemented this way (I mean, a type hinting of "string" doesn't seem to mean "string|Stringable")

But I'm not really sure, it would require the advice of an expert.

Well, wait a minute... I just saw the author of he RFC for Stringable is nobody else than you...
So it seems you're an expert :)

It's up to you to see : if I'm wrong, and mb_ functions accept string|Stringable although their prototype says "string", then you're right, and I can remove the cast.

nicolas-grekas replied:
> mb_ functions accept string|Stringable although their prototype says "string"

Yes, that works, because we don't use strict mode (and we don't want to).

derrabus replied:
> mb_ functions accept string|Stringable although their prototype says "string"

They do, unless you enable `strict_types`. 🙃

Yopai replied:
Okay, then.
Fixed

nicolas-grekas commented on src/Symfony/Component/ErrorHandler/Resources/views/exception.html.php:
```diff
@@ -16,7 +16,7 @@
     <div class="exception-message-wrapper">
         <div class="container">
-            <h1 class="break-long-words exception-message<?= mb_strlen($exceptionMessage) > 180 ? ' long' : ''; ?>"><?= $this->formatFileFromText(nl2br($exceptionMessage)); ?></h1>
+            <h1 class="break-long-words exception-message<?= mb_strlen((string) $exceptionMessage) > 180 ? ' long' : ''; ?>"><?= $this->formatFileFromText(nl2br($exceptionMessage)); ?></h1>
```
can be removed also, the exception message cannot be null

Yopai replied:
done

nicolas-grekas commented:
Thank you @Yopai.
